File: bindings/csharp/Serialization/SerializationUtils.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Linq;
using System.Reflection;

namespace MikanXR
{
	internal class ValueAccessor
	{
		public ValueAccessor(object instance, Type type)
		{
			_instance = instance;
			_type = type;
			_name = type.Name;
		}

		public ValueAccessor(object instance, FieldInfo field)
		{
			_instance = instance;
			_fieldInfo = field;
			_type = field.FieldType;
			_name = field.Name;
		}

		public void ensureValueAllocated()
		{
			if (_fieldInfo != null)
			{
				// See if the field doesn't have an instance allocated to it
				object fieldValue= _fieldInfo.GetValue(_instance);
				if (fieldValue == null)
				{
					// Allocate an instance of the field type and assign it
					fieldValue = Activator.CreateInstance(_type);
					_fieldInfo.SetValue(_instance, fieldValue);
				}
			}
			else if (_instance == null)
			{
				_instance = Activator.CreateInstance(_type);
			}
		}

		public object getValueObject()
		{
			Debug.Assert(_instance != null);

			if (_fieldInfo != null)
			{
				return _fieldInfo.GetValue(_instance);
			}
			else
			{
				return _instance;
			}
		}

		public void setValueObject(object value)
		{
			Debug.Assert(_type == value.GetType());

			if (_fieldInfo != null)
			{
				Debug.Assert(_instance != null);
				_fieldInfo.SetValue(_instance, value);
			}
			else
			{
				_instance = value;
			}
		}

		public virtual T getValue<T>()
		{
			Debug.Assert(_type == typeof(T));
			Debug.Assert(_instance != null);
			
			if (_fieldInfo != null)
			{
				return (T)_fieldInfo.GetValue(_instance);
			}
			else
			{
				return (T)_instance;
			}
		}

		public virtual void setValue<T>(T value)
		{
			Debug.Assert(_type == typeof(T));

			if (_fieldInfo != null)
			{
				Debug.Assert(_instance != null);
				_fieldInfo.SetValue(_instance, value);
			}
			else
			{
				_instance= value;
			}
		}

		protected object _instance;
		public object ValueInstance => _instance;

		protected FieldInfo _fieldInfo;
		public FieldInfo ValueField => _fieldInfo;

		protected Type _type;
		public Type ValueType => _type;

		protected string _name;
		public string ValueName => _name;
	}

	internal interface IVisitor
	{
		void visitClass(ValueAccessor accessor);
		void visitList(ValueAccessor accessor);
		void visitDictionary(ValueAccessor accessor);
		void visitPolymorphicObject(ValueAccessor accessor);
		void visitEnum(ValueAccessor accessor);
		void visitBool(ValueAccessor accessor);
		void visitByte(ValueAccessor accessor);
		void visitUByte(ValueAccessor accessor);
		void visitShort(ValueAccessor accessor);
		void visitUShort(ValueAccessor accessor);
		void visitInt(ValueAccessor accessor);
		void visitUInt(ValueAccessor accessor);
		void visitLong(ValueAccessor accessor);
		void visitULong(ValueAccessor accessor);
		void visitFloat(ValueAccessor accessor);
		void visitDouble(ValueAccessor accessor);
		void visitString(ValueAccessor accessor);
	}

	internal static class Utils
	{
		public static long getMikanClassId(Type classType)
		{
			FieldInfo classIdField= classType.GetField("classId", BindingFlags.Public | BindingFlags.Static);

			return (long)classIdField.GetValue(null);
		}

		public static void memoryOffsetSortStructFields(Type classType, ref List<FieldInfo> outFields)
		{
			// Add fields from the parent classes first
			if (classType.BaseType != null)
			{
				memoryOffsetSortStructFields(classType.BaseType, ref outFields);
			}

			// Add fields from the current class
			// TODO: Observationally, Type.GetFields() returns fields in declaration order, but it's not guaranteed
			// We should really be using a custom attribute to specify the order and then sort on that.
			// See "Remarks" in https://learn.microsoft.com/en-us/dotnet/api/system.type.getfields?view=net-9.0&redirectedfrom=MSDN#System_Type_GetFields
			FieldInfo[] fields = classType.GetFields(BindingFlags.Instance | BindingFlags.Public | BindingFlags.DeclaredOnly);
			foreach (FieldInfo field in fields)
			{
				outFields.Add(field);
			}
		}

		public static void visitObject<T>(T instance, IVisitor visitor) where T : struct
		{
			visitObject(instance, typeof(T), visitor);
		}

		public static void visitObject(object instance, Type classType, IVisitor visitor)
		{
			List<FieldInfo> fields = new List<FieldInfo>();
			memoryOffsetSortStructFields(classType, ref fields);

			foreach (FieldInfo field in fields)
			{
				visitField(instance, field, visitor);
			}
		}

		public static void visitField(object instance, FieldInfo fieldInfo, IVisitor visitor)
		{
			visitValue(new ValueAccessor(instance, fieldInfo), visitor);
		}

		public static void visitValue(ValueAccessor accessor, IVisitor visitor)
		{
			Type type = accessor.ValueType;
			string name = accessor.ValueName;

			if (type.IsClass)
			{
				Type fieldType = accessor.ValueType;

				if (fieldType.IsGenericType)
				{
					if (fieldType.Name.StartsWith("List") &&
						fieldType.GenericTypeArguments.Length == 1)
					{
						visitor.visitList(accessor);
					}
					else if (fieldType.Name.StartsWith("Dictionary") &&
							fieldType.GenericTypeArguments.Length == 2)
					{
						visitor.visitDictionary(accessor);
					}
					else if (fieldType.Name.StartsWith("SerializableObject") &&
							fieldType.GenericTypeArguments.Length == 1)
					{
						visitor.visitPolymorphicObject(accessor);
					}
					else
					{
						visitor.visitClass(accessor);
					}
				}
				else if (fieldType == typeof(string))
				{
					visitor.visitString(accessor);
				}
				else
				{
					visitor.visitClass(accessor);
				}
			}
			else if (type.IsEnum)
			{
				visitor.visitEnum(accessor);
			}
			else if (type.IsPrimitive)
			{
				switch (Type.GetTypeCode(type))
				{
					case TypeCode.Boolean:
						visitor.visitBool(accessor);
						break;
					case TypeCode.Byte:
						visitor.visitUByte(accessor);
						break;
					case TypeCode.SByte:
						visitor.visitByte(accessor);
						break;
					case TypeCode.Int16:
						visitor.visitShort(accessor);
						break;
					case TypeCode.UInt16:
						visitor.visitUShort(accessor);
						break;
					case TypeCode.Int32:
						visitor.visitInt(accessor);
						break;
					case TypeCode.UInt32:
						visitor.visitUInt(accessor);
						break;
					case TypeCode.Int64:
						visitor.visitLong(accessor);
						break;
					case TypeCode.UInt64:
						visitor.visitULong(accessor);
						break;
					case TypeCode.Single:
						visitor.visitFloat(accessor);
						break;
					case TypeCode.Double:
						visitor.visitDouble(accessor);
						break;
					default:
						throw new NotImplementedException("Accessor " + name + " has unsupported type");
				}
			}
			else
			{
				throw new NotImplementedException("Accessor " + name + " has unsupported type");
			}
		}
	}
}

File: bindings/typescript/Serialization/SerializationUtils.ts
import { PolymorphicObject, PolymorphicStruct } from '../PolymorphicObject';

/**
 * Represents a value accessor that can get/set values on objects
 * Similar to C# ValueAccessor
 */
export class ValueAccessor {
  private instance: any;
  private fieldName: string | null;
  private type: any;
  private typeName: string;

  constructor(instance: any, type: any, fieldName: string | null = null) {
    this.instance = instance;
    this.fieldName = fieldName;
    this.type = type;
    this.typeName = type?.name || 'unknown';
  }

  get valueField(): string | null {
    return this.fieldName;
  }

  get valueType(): any {
    return this.type;
  }

  get valueName(): string {
    return this.fieldName || this.typeName;
  }

  get valueInstance(): any {
    return this.instance;
  }

  ensureValueAllocated(): void {
    if (this.fieldName && !this.instance[this.fieldName]) {
      this.instance[this.fieldName] = new this.type();
    } else if (!this.instance) {
      this.instance = new this.type();
    }
  }

  getValueObject(): any {
    if (this.fieldName) {
      return this.instance[this.fieldName];
    }
    return this.instance;
  }

  setValueObject(value: any): void {
    if (this.fieldName) {
      this.instance[this.fieldName] = value;
    } else {
      this.instance = value;
    }
  }

  getValue<T>(): T {
    if (this.fieldName) {
      return this.instance[this.fieldName] as T;
    }
    return this.instance as T;
  }

  setValue<T>(value: T): void {
    if (this.fieldName) {
      this.instance[this.fieldName] = value;
    } else {
      this.instance = value;
    }
  }
}

/**
 * Visitor interface for serialization/deserialization
 */
export interface IVisitor {
  visitClass(accessor: ValueAccessor): void;
  visitList(accessor: ValueAccessor): void;
  visitDictionary(accessor: ValueAccessor): void;
  visitPolymorphicObject(accessor: ValueAccessor): void;
  visitEnum(accessor: ValueAccessor): void;
  visitBool(accessor: ValueAccessor): void;
  visitByte(accessor: ValueAccessor): void;
  visitUByte(accessor: ValueAccessor): void;
  visitShort(accessor: ValueAccessor): void;
  visitUShort(accessor: ValueAccessor): void;
  visitInt(accessor: ValueAccessor): void;
  visitUInt(accessor: ValueAccessor): void;
  visitLong(accessor: ValueAccessor): void;
  visitULong(accessor: ValueAccessor): void;
  visitFloat(accessor: ValueAccessor): void;
  visitDouble(accessor: ValueAccessor): void;
  visitString(accessor: ValueAccessor): void;
}

/**
 * Metadata for serializable fields
 */
export interface FieldMetadata {
  name: string;
  type: string;
  isArray?: boolean;
  isMap?: boolean;
  elementType?: string;
  keyType?: string;
  valueType?: string;
}

/**
 * Get serialization metadata for a class, walking the prototype chain so that
 * inherited fields from parent classes are included (parent fields first).
 */
export function getSerializationMetadata(instance: any): FieldMetadata[] {
  if (!instance || !instance.constructor) {
    return [];
  }

  // Collect metadata arrays from each class in the hierarchy, stopping at Object.
  // We push each layer to the front so parent class fields come first.
  const chain: FieldMetadata[][] = [];
  let ctor = instance.constructor;
  while (ctor && ctor !== Object) {
    if (Object.prototype.hasOwnProperty.call(ctor, '__serializationMetadata')) {
      chain.unshift(ctor.__serializationMetadata);
    }
    ctor = Object.getPrototypeOf(ctor);
  }

  return chain.flat();
}

/**
 * Visit an object and all its fields
 */
export function visitObject(instance: any, instanceType: any, visitor: IVisitor): void {
  const metadata = getSerializationMetadata(instance);

  for (const field of metadata) {
    const accessor = new ValueAccessor(instance, instanceType, field.name);
    // Store the field metadata on the accessor for use by visitors
    (accessor as any).fieldMetadata = field;
    visitField(accessor, field, visitor);
  }
}

/**
 * Visit a single field
 */
export function visitField(accessor: ValueAccessor, field: FieldMetadata, visitor: IVisitor): void {
  visitValue(accessor, field, visitor);
}

/**
 * Visit a value based on its type
 */
export function visitValue(accessor: ValueAccessor, field: FieldMetadata, visitor: IVisitor): void {
  const typeName = field.type;

  if (field.isArray) {
    visitor.visitList(accessor);
  } else if (field.isMap) {
    visitor.visitDictionary(accessor);
  } else if (typeName === 'PolymorphicObject') {
    visitor.visitPolymorphicObject(accessor);
  } else if (typeName === 'string') {
    visitor.visitString(accessor);
  } else if (typeName === 'boolean') {
    visitor.visitBool(accessor);
  } else if (typeName === 'int8') {
    visitor.visitByte(accessor);
  } else if (typeName === 'uint8') {
    visitor.visitUByte(accessor);
  } else if (typeName === 'int16') {
    visitor.visitShort(accessor);
  } else if (typeName === 'uint16') {
    visitor.visitUShort(accessor);
  } else if (typeName === 'int32' || typeName === 'number') {
    visitor.visitInt(accessor);
  } else if (typeName === 'uint32') {
    visitor.visitUInt(accessor);
  } else if (typeName === 'int64' || typeName === 'bigint') {
    visitor.visitLong(accessor);
  } else if (typeName === 'uint64') {
    visitor.visitULong(accessor);
  } else if (typeName === 'float') {
    visitor.visitFloat(accessor);
  } else if (typeName === 'double') {
    visitor.visitDouble(accessor);
  } else if (typeName.startsWith('enum:')) {
    visitor.visitEnum(accessor);
  } else {
    // Assume it's a class/struct
    visitor.visitClass(accessor);
  }
}

/**
 * Helper to get the class ID from a type
 */
export function getMikanClassId(type: any): bigint {
  if (type && type.classId !== undefined) {
    return type.classId;
  }
  return 0n;
}
